### ISSUE TITLE
When Pass experimental_relax_shapes to instance methods decorated with `tf.function`, it can't work.

### ISSUE THREAD

**Issue by wavce** [tensorflow/tensorflow#35303]
### System information

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian Stretch
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1.0rc1
- Python version: 3.7.1

### Describe the current behavior
Consider the following source:
```
import tensorflow as tf
@tf.function(experimental_relax_shapes=True)
def fn(x):
    print(tf.shape(x))
    return x * x
```
1:
```
fn(tf.constant([2]))
```
Output:
Tensor("Shape:0", shape=(1,), dtype=int32)
Out[9]: <tf.Tensor: shape=(1,), dtype=int32, numpy=array([4], dtype=int32)>

2:
```
fn(tf.constant([2, 3]))
```
Output:
Tensor("Shape:0", shape=(1,), dtype=int32)
Out[10]: <tf.Tensor: shape=(2,), dtype=int32, numpy=array([4, 9], dtype=int32)>

### Describe the expected behavior
The experimental_relax_shapes should work also on instance method.

**Comment by foxik** [comment]
I am not sure what problem you are reporting; but the problem of `experimental_relax_shapes` not working on instance methods has been solved in #35021.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the [Gist](https://colab.sandbox.google.com/gist/amahendrakar/24e22fe7e91e68780650634f2cad2aa6/35303.ipynb) here. Thanks!

**Comment by mdanatg** [comment]
Note that `tf.shape` is dynamic and always returns a `Tensor` that will contain the actual shape at runtime. What you want in this case is the static shape: `print(x.shape)`.

The logic that `experimental_relax_shapes` uses is a bit convoluted and [hard to follow](https://cs.corp.google.com/aosp-master/external/tensorflow/tensorflow/python/eager/function.py?type=cs&q=_experimental_relax_shapes&g=0&l=2131), but roughly it will only attempt to relax shapes on the third call. In general, if you enable it you should expect the function to be still called a few times times and the shapes of tensors to be anything.

1:
```
fn(tf.constant([2]))
```

Output:
(1,)
<tf.Tensor: shape=(1,), dtype=int32, numpy=array([4], dtype=int32)>

2:
```
fn(tf.constant([2, 3]))
```
Output:
(2,)
<tf.Tensor: shape=(2,), dtype=int32, numpy=array([4, 9], dtype=int32)>

3:
```
fn(tf.constant([2, 3, 4]))
```
Output:
(None,)
<tf.Tensor: shape=(3,), dtype=int32, numpy=array([ 4,  9, 16], dtype=int32)>
